src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,302 @@
+#include "firmware/app/control/trajectory_planner.h"
+
+#include <assert.h>
+#include <stdbool.h>
+
+#include "firmware/shared/math/polynomial_1d.h"
+#include "firmware/shared/math/polynomial_2d.h"
+#include "firmware/shared/math/vector_2d.h"
+#include "math.h"
+#include "stdio.h"
+
+void app_trajectory_planner_generate_constant_arc_length_segmentation(
+    FirmwareRobotPathParameters_t path_parameters, Trajectory_t* trajectory)
+{
+    float max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    float velocity_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+
+
+    // Set the internal path parameter variables
+    TrajectoryElement_t* traj_elements = trajectory->trajectory_elements;
+
+    float final_speed                      = path_parameters.final_speed;
+    float initial_speed                    = path_parameters.initial_speed;
+    const float max_allowable_acceleration = path_parameters.max_allowable_acceleration;
+    const float max_allowable_speed        = path_parameters.max_allowable_speed;
+    const unsigned int num_segments        = path_parameters.num_segments;
+    const Polynomial2dOrder3_t path        = path_parameters.path;
+    float t_end                            = path_parameters.t_end;
+    float t_start                          = path_parameters.t_start;
+
+    trajectory->num_elements = path_parameters.num_segments;
+
+    // Variable used to flag if the path is moving "backwards" along the input path
+    bool reverse_parameterization = false;
+
+    // Check that the pre conditions are met
+    assert(num_segments > 2);
+    assert(max_allowable_acceleration > 0);
+    assert(max_allowable_speed > 0);
+    assert(initial_speed >= 0);
+    assert(final_speed >= 0);
+
+    // Check for the parameterization direction
+    // If the path is traversed in reverse, then flip all components to forwards (to be
+    // reversed again in the end)
+    if (t_end < t_start)
+    {
+        reverse_parameterization = true;
+
+        // Reverse the direction (Polynomial library can only handle forwards direction)
+        float temp = t_start;
+        t_start    = t_end;
+        t_end      = temp;
+
+        // Reverse the final and initial speeds
+        temp          = initial_speed;
+        initial_speed = final_speed;
+        final_speed   = temp;
+    }
+
+    // Create the parmeterization to contain the desired number of segments
+    CREATE_STATIC_ARC_LENGTH_PARAMETRIZATION(arc_length_param,
```

Comment:
let's rename `arc_length_param` -> `arc_length_parametrization` to remove any ambiguity

Reply:
Changed